File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/CallViewController/CallViewController.m
//
//  CallViewController.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "CallViewController.h"
#import "StatsView.h"
#import "CallAction.h"
#import "ActionButton.h"

@interface CallViewController()
//MARK: - Properties
@property (strong, nonatomic) StatsView *statsView;

@end

@implementation CallViewController
//MARK - Setup
- (void)setupWithCallId:(NSString *)callId
                members:(NSDictionary<NSNumber *, NSString *>*)members
          mediaListener:(MediaListener *)mediaListener
        mediaController:(MediaController *)mediaController
              direction:(CallDirection)direction {
    self.callInfo = [CallInfo callInfoWithCallID:callId  members:members direction:direction];
    self.mediaListener = mediaListener;
    self.mediaController = mediaController;
    
    __weak __typeof(self)weakSelf = self;
    [self.mediaListener setOnAudio:^(BOOL enabled) {
        [weakSelf.actionsBar select:!enabled type:CallActionAudio];
    }];
}

- (void)endCall {
    [self.actionsBar clear];
    [self.mediaController clear];
    [self.callInfo clear];
    [self.navigationController popViewControllerAnimated:NO];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self checkCallPermissionsWithConferenceType:QBRTCConferenceTypeAudio completion:nil];
    
    __weak __typeof(self)weakSelf = self;
    NSMutableArray *callActions = @[].mutableCopy;
    
    [callActions addObject:[[CallAction alloc] initWithType:CallActionAudio action:^(ActionButton * _Nonnull sender) {
        weakSelf.mediaController.audioEnabled = !weakSelf.mediaController.audioEnabled;
    }]];
    [callActions addObject:[[CallAction alloc] initWithType:CallActionDecline action:^(ActionButton * _Nonnull sender) {
        sender.enabled = NO;
        if (weakSelf.hangUp) {
            weakSelf.hangUp(weakSelf.callInfo.callId);
        }
    }]];
    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPhone) {
        [callActions addObject:[[CallAction alloc] initWithType:CallActionSpeaker action:^(ActionButton * _Nonnull sender) {
            weakSelf.mediaController.currentAudioOutput = sender.pressed ? AVAudioSessionPortOverrideNone : AVAudioSessionPortOverrideSpeaker;
        }]];
    }
    [self.actionsBar setupWithActions:callActions.copy];
    BOOL selectedState = self.mediaController.currentAudioOutput == AVAudioSessionPortOverrideSpeaker;
    [self.actionsBar select:selectedState type:CallActionSpeaker];
    [self.actionsBar select:!self.mediaController.audioEnabled type:CallActionAudio];
    
    [self.participantsView setupWithCallInfo:self.callInfo conferenceType:QBRTCConferenceTypeAudio];
    
    
    [self.callInfo setOnChangedState:^(CallParticipant * _Nonnull participant) {
        [weakSelf.participantsView setupConnectionState: participant.connectionState participantId:participant.id];
        
        if (weakSelf.callTimer.isActive == NO && participant.connectionState == QBRTCConnectionStateConnected) {
            weakSelf.callTimer.isActive = YES;
            weakSelf.statsButton.enabled = YES;
            weakSelf.statsButton.alpha = 1.0f;
            
            if (weakSelf.callInfo.direction == CallDirectionOutgoing) {
                BOOL isPressed = [weakSelf.actionsBar isSelected:CallActionSpeaker];
                AVAudioSessionPortOverride audioPort =
                isPressed ? AVAudioSessionPortOverrideSpeaker : AVAudioSessionPortOverrideNone;
                QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
                [audioSession overrideOutputAudioPort:audioPort];
            }
        }
    }];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self.navigationController setNavigationBarHidden:YES animated:NO];
}

//MARK: - Public Methods
- (IBAction)didTapStatsButton:(UIButton *)sender {
    self.statsView = [[NSBundle mainBundle] loadNibNamed:@"StatsView" owner:nil options:nil].firstObject;
    self.statsView.callInfo = self.callInfo;
    [self.view addSubview:self.statsView];
    self.statsView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.statsView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [self.statsView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.statsView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor constant: -3.0f].active = YES;
    [self.statsView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
}

- (void)checkCallPermissionsWithConferenceType:(QBRTCConferenceType)conferenceType completion:(СheckPermissionsCompletion _Nullable)completion {
    __weak __typeof(self)weakSelf = self;
    [CallPermissions checkPermissionsWithConferenceType:QBRTCConferenceTypeAudio presentingViewController:self completion:^(BOOL granted) {
        if (granted == NO) {
            [weakSelf.actionsBar select:YES type:CallActionAudio];
            [weakSelf.actionsBar setUserInteractionEnabled:NO type:CallActionAudio];
            weakSelf.mediaController.audioEnabled = NO;
        }
        if (conferenceType == QBRTCConferenceTypeAudio) {
            if (completion) {
                completion(granted);
            }
            return;
        }
    }];
    
    if (conferenceType == QBRTCConferenceTypeVideo) {
        [CallPermissions checkPermissionsWithConferenceType:QBRTCConferenceTypeVideo
                                   presentingViewController:self completion:^(BOOL granted) {
            if (granted == NO) {
                [weakSelf.actionsBar select:YES type:CallActionVideo];
                [weakSelf.actionsBar select:YES type:CallActionSwitchCamera];
                [weakSelf.actionsBar setUserInteractionEnabled:NO type:CallActionVideo];
                [weakSelf.actionsBar setUserInteractionEnabled:NO type:CallActionSwitchCamera];
            }
            weakSelf.mediaController.videoEnabled = granted;
            if (completion) {
                completion(granted);
            }
        }];
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/AudioCallViewController/CallViewController.swift
//
//  AudioCallViewController.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 27.04.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import Foundation
import UIKit
import QuickbloxWebRTC

class CallViewController: UIViewController {
    //MARK: - IBOutlets
    @IBOutlet weak var headerView: CallGradientView!
    @IBOutlet weak var bottomView: CallGradientView!
    @IBOutlet weak var actionsBar: CallActionsBar! {
        didSet {
            var actionButtons: [(ActionType, action: (_ sender: ActionButton?) -> Void)] = []
            actionButtons.append((.audio, action: { [weak self] sender in
                guard let self = self else {
                    return
                }
                self.mediaController.audioEnabled = !self.mediaController.audioEnabled
            }))
            
            actionButtons.append((.decline, action: { [weak self] sender in
                guard let self = self else {
                    return
                }
                sender?.isEnabled = false
                self.hangUp?(self.callInfo.callId)
            }))
            
            if UIDevice.current.userInterfaceIdiom == .phone {
                actionButtons.append((.speaker, action: { sender in
                    guard let speakerButton = sender else { return}
                    let audioSession = QBRTCAudioSession.instance()
                    if audioSession.isActive == false {
                        return
                    }
                    let audioPort: AVAudioSession.PortOverride = speakerButton.pressed == false ? .speaker : .none
                    audioSession.overrideOutputAudioPort(audioPort)
                }))
            }
            actionsBar.setup(withActions: actionButtons)
            
            let selectedState = mediaController.currentAudioOutput == AVAudioSession.PortOverride.speaker
            actionsBar.select(selectedState, type: .speaker)
            actionsBar.select(!mediaController.audioEnabled, type: .audio)
        }
    }
    @IBOutlet weak var participantsView: ParticipantsView!
    
    @IBOutlet weak var statsButton: UIButton! {
        didSet {
            statsButton.isEnabled = false
            statsButton.alpha = 0.0
        }
    }
    
    @IBOutlet weak var callTimer: CallTimerView!
    
    lazy var statsView: StatsView = {
        let statsView = StatsView.loadNib()
        return statsView
    }()
    
    //MARK: - Properties
    var callInfo: CallInfo!
    var hangUp: CallHangUpAction?
    var mediaListener: MediaListener!
    var mediaController: MediaController!
    
    //MARK - Setup
    func setupWithCallId(_ callId: String, members: [NSNumber: String], mediaListener: MediaListener, mediaController: MediaController, direction: CallDirection) {
        callInfo = CallInfo(callId: callId, members: members, direction: direction)
        self.mediaListener = mediaListener
        self.mediaController = mediaController
        
        self.mediaListener.onAudio = { [weak self] enable in
            if self?.actionsBar == nil {
                return
            }
            self?.actionsBar.select(!enable, type: .audio)
        }
    }
    
    func setupViews() {
        participantsView.setup(callInfo: callInfo, conferenceType: .audio)
        checkCallPermissions(.audio, completion: nil)

        callInfo.onChangedState = { [weak self] (participant) in
            self?.participantsView.setConnectionState(participant.connectionState, participantId: participant.id)
            
            if self?.callTimer.isActive == false, participant.connectionState == .connected {
                self?.callTimer.isActive = true
                self?.statsButton.isEnabled = true
                self?.statsButton.alpha = 1.0
                
                if self?.callInfo.direction == .outgoing,
                   let isPressed = self?.actionsBar.isSelected(.speaker) {
                    let audioPort: AVAudioSession.PortOverride = isPressed ? .speaker : .none
                    let audioSession = QBRTCAudioSession.instance()
                    audioSession.overrideOutputAudioPort(audioPort)
                }
            }
        }
    }
    
    //MARK: - Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupViews()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        navigationController?.setNavigationBarHidden(true, animated: animated)
    }
    
    @IBAction func didTapStatsButton(_ sender: UIButton) {
        statsView.callInfo = callInfo
        
        view.addSubview(statsView)
        statsView.translatesAutoresizingMaskIntoConstraints = false
        statsView.rightAnchor.constraint(equalTo: view.rightAnchor, constant: 0.0).isActive = true
        statsView.topAnchor.constraint(equalTo: view.topAnchor, constant: 0.0).isActive = true
        statsView.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 0.0).isActive = true
        statsView.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: 0.0).isActive = true
    }
    
    func checkCallPermissions(_ conferenceType: QBRTCConferenceType, completion:((_ videoGranted: Bool) -> Void)?) {
        CallPermissions.check(with: .audio, presentingViewController: self) { [weak self] audioGranted in
            if audioGranted == false {
                self?.mediaController.audioEnabled = false
                self?.actionsBar.select(true, type: .audio)
                self?.actionsBar.setUserInteractionEnabled(false, type: .audio)
            }
            if conferenceType == .audio {
                completion?(audioGranted)
                return
            }
        }
        
        if conferenceType == .video {
            CallPermissions.check(with: .video, presentingViewController: self) { [weak self] videoGranted in
                if videoGranted == false {
                    self?.actionsBar.select(true, type: .video)
                    self?.actionsBar.select(true, type: .switchCamera)
                    self?.actionsBar.setUserInteractionEnabled(false, type: .video)
                    self?.actionsBar.setUserInteractionEnabled(false, type: .switchCamera)
                }
                self?.mediaController.videoEnabled = videoGranted
                completion?(videoGranted)
            }
        }
    }
}
